04-onion-routing.md

```diff
@@ -238,11 +236,47 @@ Field descriptions:
      leaking its position in the route.
 
    * `padding`: This field is for future use and also for ensuring that future non-0-`realm`
-     `per_hop`s won't change the overall `hops_data` size.
+     `hop_data`s won't change the overall `hop_payloads` size.
+
+When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified
+within `hop_data` above; otherwise, deviation from the specified HTLC
+parameters may lead to extraneous routing failure.
+
+### `tlv_payload` payload format
+
+This is a more flexible format, which avoids the redundant `short_channel_id` field for the final node. 
+
+1. tlv: `tlv_payload`
+2. types:
+    1. type: 0 (`destination_signal`)
+    2. data:
+	    * Empty
+    1. type: 2 (`amt_to_forward`)
+    2. data:
+        * [`integer`:`amt_to_forward`]
+    1. type: 4 (`outgoing_cltv_value`)
+    2. data:
+        * [`integer`:`outgoing_cltv_value`]
+    1. type: 6 (`short_channel_id`)
+    2. data:
+        * [`8`:`short_channel_id`]
```

Comment:
I'd like to bring up an item from the 2018 Lightning Summit: Should we use the introduction of a new onion format as an opportunity to switch to next hop pubkeys instead of short channel ids? Mentioned advantages:
* Greater deniability for unadvertised nodes, can use ephemeral pubkeys when connecting to “gateway” (edge) nodes
* Possibly opens up to allow onion overlay (not sure what that is)

And also:

* Shorter hop hints for private channels, as they only need to include a pubkey.
* Possible implementation advantages because of simplification of data structures (like for example `route`).

Reply:
While this would be a great addition, I feel that it should be a separate change. The current PR is about the encoding format of payloads, not really about their contents and routing logic (and this PR is already a big change and adding a new change like that would trigger many new discussions that are irrelevant to the variable-length payload).

I think it would be cleaner to use a dedicated feature-flag (something like `pubkey_onion_routing`) separate from the `var_onion_optin` flag. Or bundle this with a few other onion tlvs that would enable multiple routing improvements?

A small drawback is that a `public_key` uses more bytes than a `short_channel_id`, so I think that nodes would support both options.

Reply:
The reason for proposing it here, is to explore whether we see an advantage in not having both options.

Reply:
I think it would be limiting to completely drop support of `short_channel_id` in favor of `public_key`. In cases where we'll want to put more information in the onion (rendezvous, trampoline, stuckless payments, etc), space will be scarce if we still want to guarantee a good anonymity set (via a potentially long route) so it's useful to have something shorter than a `public_key` to identify the next hop.

Reply:
I think it's probably a good idea to have that discussion in its own PR like @t-bast suggested. I already removed the requirement for there to be a `short_channel_id` (the text was forcing us to error out if no `short_channel_id` is present so far, see last commit). This allows us to then simply switch out the `short_channel_id` for any other form of addressing.

One option for addressing would also be to have a small local peer ID, i.e., a node announces a small integer ID for each of its peers that can be used as an alias when routing. The reason being that this alias only has to make sense to the processing node, hence we can use small integers. More on this in a separate discussion.

I just feel this change has been hanging in limbo for far too long, despite everybody agreeing that it is something that we desperately need (this is the 3rd iteration of this general concept), and there are quite a few interesting things that are being blocked by this. Let's not add more details that can be trivially deferred to later.

Reply:
Fair points. I just wanted to check whether we don't see it as a missed opportunity to hard switch to a different addressing mode in this new onion format. As it is clear that pubkey addressing isn't something that we unanimously want as the only addressing mode, this can indeed be deferred to later.